### ISSUE TITLE
Getting AccessViolationException while adding custom resource dictionary in App.Xaml file in Release builds

### ISSUE THREAD

**Issue by naveenkumar-sanjeevirayan** [dotnet/maui#19535]
### Description

I have created a Maui library project file to add a custom resource dictionary for theme support. Then, I included it in the application file and added the custom resource dictionary to the merged dictionaries like below. I'm getting AccessViolationException when I run the application in release mode only. And the example deployed successfully with debug mode. I have attached the runnable example with the library project.

```
<Application xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             xmlns:local="clr-namespace:CustomThemeTesting"
             xmlns:theme="clr-namespace:ThemeLibrary;assembly=ThemeLibrary"
             x:Class="CustomThemeTesting.App">
    <Application.Resources>
        <ResourceDictionary>
            <ResourceDictionary.MergedDictionaries>
                <ResourceDictionary Source="Resources/Styles/Colors.xaml" />
                <ResourceDictionary Source="Resources/Styles/Styles.xaml" />
                <theme:CustomThemeResourceDictionary VisualTheme="LightDefault"/>
            </ResourceDictionary.MergedDictionaries>
        </ResourceDictionary>
    </Application.Resources>
</Application>
```

Error message: 
<img width="317" alt="AccessViolationException" src="https://github.com/dotnet/maui/assets/61832185/27229d91-ad72-4273-9062-75910ea58ce3">

### Steps to Reproduce

1. Download and extract the attached example.
2. Run the CustomThemeTesting project with release mode. You will get the access violation exception.

[Maui_Theme_Issue.zip](https://github.com/dotnet/maui/files/13737233/Maui_Theme_Issue.zip)

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android, Windows, macOS

**Comment by PureWeen** [comment]
@naveenkumar-sanjeevirayan are you seeing this on all 4 platforms? Just want to make sure you made an accurate selection there.

**Comment by naveenkumar-sanjeevirayan** [comment]
Hi, @PureWeen I'm facing this on Windows and Android platforms. I didn't check in Mac and Ios. So, Please ensure from your side for Mac and iOS platforms.

**Comment by rlennoxmemoro** [comment]
> @naveenkumar-sanjeevirayan are you seeing this on all 4 platforms? Just want to make sure you made an accurate selection there.

Hi - I actually reported this to SyncFusion, and they opened the bug.

It failed on iOS, Windows, and Android. I don't use it in MacCatalyst, so I don't know about that one.

If you simply download and run the sample you can see it though.

**Comment by PaulAndersonS** [comment]
This works fine in .NET7.0 and issue occurring in .NET8.0

**Comment by PaulAndersonS** [comment]
@PureWeen  Can you check on this

**Comment by Ahamed-Ali** [comment]
Any update on this?... Please resolve this as soon as possible or any workaround available ?

**Comment by StephaneDelcroix** [comment]
trying to reproduce... do you have a call stack ? what happens if you disable Xaml Compilation on App.xaml ?

**Comment by StephaneDelcroix** [comment]
We found that the IL generated for your custom themes inheriting from CustomeThemeDictionary is wrong. you can disable Xaml compilation on those (by attributing the code behind class with the `[XamlCompilation(XamlCompilationOptions.Skip)]` attribute as a workaround

**Comment by StephaneDelcroix** [comment]
also, this code 
<img width="414" alt="image" src="https://github.com/dotnet/maui/assets/313003/0f3186ee-9cf7-4914-9db3-19c2d8113a2c">
 in in CustomThemeDictionary.xaml.cs looks highly suspicious

**Comment by StephaneDelcroix** [comment]
I was unable to get the AccessViolationException, but there was indeed a NullRefException (that might have caused the AccessViolation one). there's a fix for it in #20463

**Comment by Choza-rajan** [comment]
We tested the nightly build 8.0.7 as per your suggestion but still we face the reported issue. Also, we further checked the test case you have added in your [fix PR](https://github.com/dotnet/maui/pull/20463/files), and we found that the scenario is different compared to others. We have explained that in the below picture.

![image](https://github.com/dotnet/maui/assets/92723643/45b9b08c-4a0d-4136-b9db-6885d67b19ca)

We have even simplified the replicating sample with the above scenario. In the attached sample We have created a class CustomThemeDictionary : ResourceDictionary. Then another class as LightThemeColors : CustomThemeDictionary. Keys in the LightThemeColors causes the crash whereas if the keys defined directly in CustomThemeDictionary works. Also below is the error screenshot for reference.

**Note: Issue occurs only in Release mode.**

![image002](https://github.com/dotnet/maui/assets/92723643/6cfbb737-5c1e-4d5d-bf5c-0900ed2e5fae)

Regarding the suggested workaround, it works for us skipping the XamlCompilation.

Sample link: [CustomThemeTesting 1.zip](https://github.com/dotnet/maui/files/14306201/CustomThemeTesting.1.zip)

**Comment by StephaneDelcroix** [comment]
@Choza-rajan what you're describing is exactly what the PR is for. Investigating if the PR is in the nightly you tested or not...

**Comment by StephaneDelcroix** [comment]
@Choza-rajan in order to debug or get a stack trace, you could force XamlCompilation on DEBUG builds by setting `MauiXamlCValidateOnly` to `false` in your csproj. By doing that, you won't get the tooling you usually get (hot reload, etc, ...

**Comment by StephaneDelcroix** [comment]
the fix isn't available in 8.0.7, but should be in 8.0.10-nightly.10044
